engine/runner/src/main/java/org/enso/runner/EnsoLibraryFeature.java

```diff
@@ -0,0 +1,77 @@
+package org.enso.runner;
+
+import static scala.jdk.javaapi.CollectionConverters.asJava;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashSet;
+import java.util.TreeSet;
+import org.enso.compiler.core.ir.module.scope.imports.Polyglot;
+import org.enso.pkg.PackageManager$;
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.hosted.RuntimeReflection;
+
+public final class EnsoLibraryFeature implements Feature {
+  @Override
+  public void beforeAnalysis(BeforeAnalysisAccess access) {
+    var libs = new LinkedHashSet<Path>();
+    for (var p : access.getApplicationClassPath()) {
+      var p1 = p.getParent();
+      if (p1 != null && p1.getFileName().toString().equals("java")) {
+        var p2 = p1.getParent();
+        if (p2 != null
+            && p2.getFileName().toString().equals("polyglot")
+            && p2.getParent() != null) {
+          libs.add(p2.getParent());
+        }
+      }
+    }
+
+    var classes = new TreeSet<String>();
+    try (var parser = new org.enso.compiler.core.EnsoParser()) {
+      for (var p : libs) {
+        var result = PackageManager$.MODULE$.Default().loadPackage(p.toFile());
+        if (result.isSuccess()) {
+          var pkg = result.get();
+          for (var src : pkg.listSourcesJava()) {
+            var code = Files.readString(src.file().toPath());
+            var ir = parser.compile(code);
+            for (var imp : asJava(ir.imports())) {
+              if (imp instanceof Polyglot poly && poly.entity() instanceof Polyglot.Java entity) {
+                var name = new StringBuilder(entity.getJavaName());
+                Class<?> clazz;
+                for (; ; ) {
+                  clazz = access.findClassByName(name.toString());
+                  if (clazz != null) {
+                    break;
+                  }
+                  int at = name.toString().lastIndexOf('.');
+                  if (at < 0) {
+                    throw new IllegalStateException("Cannot load " + entity.getJavaName());
+                  }
+                  name.setCharAt(at, '$');
+                }
+                classes.add(clazz.getName());
+                RuntimeReflection.register(clazz);
+                RuntimeReflection.register(clazz.getConstructors());
+                RuntimeReflection.register(clazz.getMethods());
+                RuntimeReflection.register(clazz.getFields());
+                RuntimeReflection.registerAllConstructors(clazz);
+                RuntimeReflection.registerAllFields(clazz);
+                RuntimeReflection.registerAllMethods(clazz);
+              }
+            }
+          }
+        }
+      }
+    } catch (Exception ex) {
+      ex.printStackTrace();
+      throw new IllegalStateException(ex);
+    }
+    System.err.println("Summary for polyglot import java:");
```

Comment:
For every registered class, I have to register all its `public`:
- methods
- constructors
- fields

That may be **too much** - as not all these methods are really called by Enso code. As such, it might be better to use following pattern. Rather than calling [org.w3c.dom.Element.getTagName()](https://docs.oracle.com/en/java/javase/21/docs/api/java.xml/org/w3c/dom/Element.html#getTagName()) you could wrap the call into:
```java
class XML_Utils {
  public static String getTagName(Element e) {
    return e.getTagName();
  }
}
```
that way we would _concentrate and select_ only the needed methods and the **transitive closure** of needed code computed by `native-image` might be smaller.

Reply:
> isn't that a bit too much?

Looking at the list you provided, every of these classes seems needed to provide some of the functionality of the standard library. The std-lib provides a broad variety of functionalities so it seems expected it may have some dependencies. 153 does not seem _that_ large to me.

I guess we may want to discuss whether we should split off some of these dependencies away from `Base` into separate libraries. I guess we could indeed do so, to keep a 'lighter' "core". IF we split off, I suspect most of our base workflows will for the time being import most of the separated parts anyway, as they are expected to be available for the users out of the box. So what could be the benefit of splitting?

I expect @jdunkerley may have some opinions on whether we want to keep a 'big' `Base` library or make it more modularized.

Reply:
> 153 does not seem that large to me.

We shouldn't be counting just the classes, but rather **methods**. The problem with current approach is that by using a _method rich class_ like [HttpClient](https://docs.oracle.com/en/java/javase/21/docs/api/java.net.http/java/net/http/HttpClient.html) we _"bring in"_ all its methods (including for example such a _megamorphic methods_ like `equals`, `hashCode` and `toString`), while we are probably using just a few. The

```java
class XML_Utils {
  public static String getTagName(Element e) {
    return e.getTagName();
  }
}
```
allows us to _select only the needed methods_ and the _transitive closure_ of needed code computed by `native-image` would be **smaller**. Hence I'd like you to use this _static method wrapper_ where possible for `Standard` libraries to be eligible for native image compilation.

Right now the `./runner` has 418MB (with Graal.js and GraalPython compiled in). By carefully selecting what goes into the image we might get to 410MB or maybe even below 400MB.

Reply:
> Right now the ./runner has 418MB (with Graal.js and GraalPython compiled in). By carefully selecting what goes into the image we might get to 410MB or maybe even below 400MB.

That's like 5%? Seems like a lot of hassle and maintenance for reducing the image by that amount.

Reply:
> That's like 5%? 

Yes, it is not going to be much.

> Seems like a lot of hassle and maintenance for reducing the image by that amount.

On contrary, I see that (e.g. **exactly specifying what methods are used**) as a good coding and architecture practice regardless of the expected gain in the size of native executable. Way more responsible than _importing everything_ and hoping someone makes the result is efficient.

If this native compilation of libraries gets to production, we should make sure we squeeze `Standard` libraries to make them as much effective as possible.